### ISSUE TITLE
2020.10 msi?

### ISSUE THREAD

**Issue by ToddAndMargo** [rakudo/rakudo#4001]
May we please have an msi of 2020.10?

I need the fix from https://github.com/rakudo/rakudo/issues/3582

**Comment by lkppo** [comment]
Hello,

Month after month no MSI release of Rakudo Star Bundle happens.

What is blocking?

Regards,

**Comment by patrickbkr** [comment]
Basically missing man power and a most Rakudo devs not having a Windows available.

I am currently in the process of setting up automated creation of MSIs for the monthly Rakudo releases. Maybe the Star Bundle can benefit from this work. No promises though.

**Comment by lkppo** [comment]
I would be happy to help. Whet could be done?

**Comment by patrickbkr** [comment]
@lkppo I think @AntonOks has [been working on an automation to build the Star Bundle](https://github.com/AntonOks/rakudo-star-win). I'd recommend to contact him and see how to best work towards finalizing the automation and then sharing the load of actually running the build toolchain every three months (I think that's the interval the Star Bundle aims for), verifying the resulting build and publishing it.

**Comment by lkppo** [comment]
Ok, I send a mail to @AntonOks.

**Comment by AntonOks** [comment]
Hi all.

Right now everything in my REPO work smooth and is based on the [build-with-choco.ps1](https://github.com/AntonOks/rakudo-star-win/blob/master/build-with-choco.ps1) script.
This script can be run locally on any (recent) Windows OS, which is supported by [chocolatey ](https://chocolatey.org/) _AND_ is also used in my [GitHub Action/workflow](https://github.com/AntonOks/rakudo-star-win/blob/master/.github/workflows/build_rakudo_star_win.yml). The GH workflow kicks of as soon as a (new) [TAG (!!!) or BRANCH is created / pushed](https://github.com/AntonOks/rakudo-star-win/blob/10f2993820281825f15d3ffde84b5a677a2f329e/.github/workflows/build_rakudo_star_win.yml#L5)

I tried to explain things in the README of my REPO. Seems like it's not really helpful...

  

I have no real clue how to proceed to get the MSIs into the [RAKUDO.org download section](https://rakudo.org/downloads/star). There are plenty options I can think off:

1. Someone with RAKUDO.org upload permissions just grabs the files from my REPO and copies them to wherever they are needed
2. Someone uses whatever helps from my REPO and extends https://github.com/rakudo/star so that all supported STAR packages are build here
    1. No idea how those would land in the download section, see above. But I assume there is already a way how someone is doing it today
    2. I started to work on this and intended to sent a merge request once I've integrated the _Windows_ STAR MSI package creation, but later figured out that also the Linux / RSTAR build doesn't create any package on GitHub, so consider to also change the "Linux build" now...
3. Someone uses the GitHub API to download my MSI files and extends whatever is available today to get the STAR MSI packages into [RAKUDO.org](https://rakudo.org/downloads/star)


I think, having all the various RAKUDO REPOs on GitHub today, we should build a solution, which kicks off and builds all the required packages (_ALL_ incl. STAR!) as soon as a new official RAKUDO release was published (from my understanding, this it just a new TAG, like https://github.com/rakudo/rakudo/tree/2021.06). Saying this, it would also mean _ALL_ packages could should be available on GitHub in the related "Releases" sections in the future.


Another thought... do we really need STAR in the future? As RAKUDO will soon come with an MSI _AND_ RAKUDO has already ZEF today, whats the benefit of STAR?

**Comment by patrickbkr** [comment]
The process of creating the Rakudo pre-compiled packages is automated already. The pipeline needs to be triggered manually and signing and upload to rakudo.org is a manual process as well. I actually like it that way. It gives me a chance to have a last look at the files before they go live.

Originally the idea of the Star Bundle was to have a more reliable release quarterly. The Rakudo devs back then would make extra sure the Rakudo release that would be the base of a Star Bundle would be stable. But since the quality standard of the monthly releases in general increased (thanks to our Release Managers), there is now less need for an extra stable quarterly release.

The Star Bundle contains a set of pre-compiled modules. (The list of included modules is in need of an update, but that's a separate story.) As @AntonOks already mentioned, now that we have the pre-compiled monthly releases which include the Zef module installer, there is a less pressing need for the Star Bundle. I personally am not a big fan of the idea of bundling loads of modules (I'd rather encourage people to learn how to use Zef instead), but that's my personal opinion and I know other people disagree.

We sadly lost the current Star Bundle maintainer, but [the tools to create the Star Bundle source release](https://gitlab.com/tyil/rakudo-star) are still available. I'd like to make sure we do the source release along side the MSI. If I remember correctly the Star Bundle source release does contain the pre-compiled modules, so only Rakudo itself needs to be compiled to have a functioning installation. It's worth a thought to base the MSI package on this source release instead of creating it from scratch. 

I am still working to improve the Wix files that create the actual MSI. Currently I work towards integrating that into the monthly Rakudo release process. I have my hopes up that I manage to have this ready for the 07.2021 release. Adapting these Wix files to the Star Bundle should be possible.

So the possible tasks as I see them are:

- Familiarize with the Star Bundle source release process.
- Think about if it makes sense to change the MSI build system AntonOks created to rely on the Star Bundle source release.
- Think about who is willing to commit to creating the Star Bundle every three months. (Or more often if willing.)
- (Once I'm done with the Wix files) Port these new Wix files over to the Star Bundle.

@AntonOks do you consider your 2021.04 Star Bundle MSI stable enough to publish on rakudo.org? If yes and you are ok with it, I'll push it to rakudo.org.

**Comment by patrickbkr** [comment]
Should no one be willing to invest the time to do further work on the Star Bundle, it is a valid alternative to just pause further Star Bundle releases. In that case I'd like to place a respective message on rakudo.org informing visitors.

@AntonOks @lkppo What do you think?

**Comment by AntonOks** [comment]
> 
> 
> The process of creating the Rakudo pre-compiled packages is automated already. The pipeline needs to be triggered manually and signing and upload to rakudo.org is a manual process as well. I actually like it that way. It gives me a chance to have a last look at the files before they go live.

Ok, fair enough. Good to know you are the one who copies files to RAKUDO.ORG :)

> Originally the idea of the Star Bundle was to have a more reliable release quarterly. The Rakudo devs back then would make extra sure the Rakudo release that would be the base of a Star Bundle would be stable. 

Hmmm... how did they do that? Is there a defined test suite or something similar?

> But since the quality standard of the monthly releases in general increased (thanks to our Release Managers), there is now less need for an extra stable quarterly release.


> The Star Bundle contains a set of pre-compiled modules. (The list of included modules is in need of an update, but that's a separate story.)

I guess, you talk about [the modules.txt file](https://raw.githubusercontent.com/rakudo/star/master/etc/modules.txt)?
I'm fetching [it here](https://github.com/AntonOks/rakudo-star-win/blob/10f2993820281825f15d3ffde84b5a677a2f329e/build-with-choco.ps1#L89), [iterating over the given module names and try to install the recent version of each of those via ZEF
](https://github.com/AntonOks/rakudo-star-win/blob/10f2993820281825f15d3ffde84b5a677a2f329e/build-with-choco.ps1#L96)

> As @AntonOks already mentioned, now that we have the pre-compiled monthly releases which include the Zef module installer, there is a less pressing need for the Star Bundle. I personally am not a big fan of the idea of bundling loads of modules (I'd rather encourage people to learn how to use Zef instead), but that's my personal opinion and I know other people disagree.

**My viewpoint:** The STAR MSI bundle is a great help within company networks, where usually servers and clients don't have access to the internet. MSIs are easy to be deployed and managed by IT departments _AND_ then again STAR should come with already "batteries included" for 95% of the RAKU dev's / admins needs, 

> We sadly lost the current Star Bundle maintainer, but [the tools to create the Star Bundle source release](https://gitlab.com/tyil/rakudo-star) are still available. I'd like to make sure we do the source release along side the MSI. If I remember correctly the Star Bundle source release does contain the pre-compiled modules, so only Rakudo itself needs to be compiled to have a functioning installation. It's worth a thought to base the MSI package on this source release instead of creating it from scratch.

Well, as I wrote, I think a fully automated build of all STAR bundles (Windows, Linux, macOS) could and should be done with one (worst case, still more) GitHub workflows. That's why I forked [GitHub RAKUDO STAR](https://github.com/rakudo/star), which is again also just another "mirror" of the one from GitLab you mentioned, see [repositories](https://github.com/rakudo/star#repositories), to https://github.com/AntonOks/star.git Again, as said before, initially I wanted to integrate my [rakudo-star-win](https://github.com/AntonOks/rakudo-star-win.git) build script / workflow into https://github.com/rakudo/star ... in addition I'll make sure also a "Linux STAR build, based on RSTAR" will be done in parallel to the Windows-STAR build.

> I am still working to improve the Wix files that create the actual MSI. Currently I work towards integrating that into the monthly Rakudo release process. I have my hopes up that I manage to have this ready for the 07.2021 release. Adapting these Wix files to the Star Bundle should be possible.
> 
> So the possible tasks as I see them are:
> 
>     * Familiarize with the Star Bundle source release process.
> 
>     * Think about if it makes sense to change the MSI build system AntonOks created to rely on the Star Bundle source release.
> 
>     * Think about who is willing to commit to creating the Star Bundle every three months. (Or more often if willing.)
> 
>     * (Once I'm done with the Wix files) Port these new Wix files over to the Star Bundle.
> 
> 
> @AntonOks do you consider your 2021.04 Star Bundle MSI stable enough to publish on rakudo.org? If yes and you are ok with it, I'll push it to rakudo.org.

Well, define "stable enough"? I don't run an extensive test farm ;) but so far there is only positive feedback... if any. And yes, if you ask me, **_I'd be very happy if you'd push the files to RAKUDO.ORG so users / devs would see this update_**.

PS: As everything "just works", I'll continue to build [rakudo-star-win MSI releases](https://github.com/AntonOks/rakudo-star-win/releases) whenever a new Rakudo version is published... until we have something better available

PPS: Who was building the macOS releases? Is there something available which could be used to also provide more recent macOS STAR packages, or...?

**Comment by AntonOks** [comment]
> Should no one be willing to invest the time to do further work on the Star Bundle, it is a valid alternative to just pause further Star Bundle releases. In that case I'd like to place a respective message on rakudo.org informing visitors.
> 
> @AntonOks @lkppo What do you think?

All the work to build the Windows-STAR release and make it available on GitHub is done and I can / will provide STAR MSI packages in my REPO...  every month, just after an official RAKUDO release...

**Comment by lkppo** [comment]
> My viewpoint: The STAR MSI bundle is a great help within company networks, where usually servers and clients don't have access to the internet. MSIs are easy to be deployed and managed by IT departments AND then again STAR should come with already "batteries included" for 95% of the RAKU dev's / admins needs,

I agree with @AntonOks. A bundle is great when you are behind a misconfigured/picky proxy or are the inmate of your sysadmin. But even for personal use I like to have a stable package that changes according to my schedule.

**Comment by lkppo** [comment]
@AntonOks 

> Right now everything in my REPO work smooth and is based on the build-with-choco.ps1 script.
> This script can be run locally on any (recent) Windows OS, which is supported by chocolatey AND is also used in my GitHub Action/workflow. The GH workflow kicks of as soon as a (new) TAG (!!!) or BRANCH is created / pushed
> 
> I tried to explain things in the README of my REPO. Seems like it's not really helpful...

Your README is neat. It's just that for me it was a discovery.

**Comment by lkppo** [comment]
@AntonOks

Digging the last dmg bundle for macos, the **.DS_Store** file seems to point to @rba (see https://giters.com/rba/star) as the macos release builder.

**Comment by patrickbkr** [comment]
@AntonOks
> Hmmm... how did they do that? Is there a defined test suite or something similar?

I think back then there was no formal process. It was mostly a matter of leaving enough time for testing, not introducing massively new features and really only release if there are no new known problems left.
Today there is Blin, the ecosystem tester that is run prior to each release and tracks all regressions in all ecosystem modules. Any regression is a release blocker.

> I guess, you talk about [the modules.txt file](https://raw.githubusercontent.com/rakudo/star/master/etc/modules.txt)?

Exactly.

> Again, as said before, initially I wanted to integrate my [rakudo-star-win](https://github.com/AntonOks/rakudo-star-win.git) build script / workflow into https://github.com/rakudo/star ... in addition I'll make sure also a "Linux STAR build, based on RSTAR" will be done in parallel to the Windows-STAR build.

I'm really grateful for your work on this.

Can you also see if you manage to create the Star Bundle source release?

> PPS: Who was building the macOS releases? Is there something available which could be used to also provide more recent macOS STAR packages, or...?

I haven't talked about it yet, but I planed to work on MacOS packaging after finishing up the MSI work. I'm not opposed to someone else picking this up first, though.

**Comment by patrickbkr** [comment]
@AntonOks @ToddAndMargo I just uploaded the Star Bundle 2021.04. <https://rakudo.org/star>
Thanks for your work AntonOks!

**Comment by patrickbkr** [comment]
I am now content with the MSI installer and the 2021.07 release included the Windows MSI.

@AntonOks

Would you be willing to update the Star Bundle MSI installer to incorporate all the changes in the WiX files that are now in the Rakudo MSI installer? (I think just copying it and doing some small modifications to accommodate the differences of the Rakudo and Star Bundle installers should do the trick.)

I can upload the 2021.07 Star Bundle installer you built to rakudo.org. Do you think I should do so, given that the Rakudo and Star Bundle installers are not in sync yet and the Star Bundle Source release is missing?

**Comment by AntonOks** [comment]
Hi @patrickbkr 

1. Very cool. I saw your MSI merge and also had a quick look into your WIX file... and yes, in the (near) future I'll try to implement the same "look & feel" into the Star MSI installer...
2. I'd say, please upload the current 2021.07 Star MSI. The WIX change will take some time and I think it's better to give the developers / Rakudo users a as recent as possible Rakudo Star version.
3. If you say "Star Bundle Source" you refer to the Linux RSTAR tool "src bundle", right?
  If so, I've cloned [rakudo/star](https://github.com/rakudo/star), implemented some changes in the RSTAR tool and added my rakudo-star-win repo into it. The result is [here ](https://github.com/AntonOks/star/releases )and the plan is to merge changes into [rakudo/star](https://github.com/rakudo/star) so later ONE "Star" release for all supported clients will be done via ONE GitHub actions workflow...

**Comment by patrickbkr** [comment]
@AntonOks

> 1. Very cool. I saw your MSI merge and also had a quick look into your WIX file... and yes, in the (near) future I'll try to implement the same "look & feel" into the Star MSI installer...

Just to make sure this isn't missed: I'm very convinced copying those WiX files and doing minor changes should be the way to go. Feel free to consult me, should you need help. (I've bitten through that WiX stuff, no need for you to do the same.)

> 2. I'd say, please upload the current 2021.07 Star MSI. The WIX change will take some time and I think it's better to give the developers / Rakudo users a as recent as possible Rakudo Star version.

I'll do the upload this evening.

> 3. If you say "Star Bundle Source" you refer to the Linux RSTAR tool "src bundle", right?

I refer to the rightmost link on <https://rakudo.org/star>. I think we are talking about the same thing.

> If so, I've cloned [rakudo/star](https://github.com/rakudo/star), implemented some changes in the RSTAR tool and added my rakudo-star-win repo into it. The result is [here ](https://github.com/AntonOks/star/releases)and the plan is to merge changes into [rakudo/star](https://github.com/rakudo/star) so later ONE "Star" release for all supported clients will be done via ONE GitHub actions workflow...

That sounds pretty good.

**Comment by coke** [comment]
With MSIs apparently being released again (Thank you!) , Closing this ticket that was originally about a 2020 MSI; If we need to track something more specific, let's open a new ticket with those details.

Thanks, folks!
